### PR TITLE
[IOTDB-6261] Pipe: support leader cache in iotdb-thrift-sync-connector

### PR DESCRIPTION
**Pull request by xuanronaldo**
create a sync task with leader cache:
```SQL
create pipe A2B
with sink (
  'sink'='iotdb-thrift-sync-sink',
  'sink.ip'='127.0.0.1',
  'sink.port'='16667',
  'sink.batch.enable'='false',
  'sink.leader-cache.enable'='true'
)

### CODE REVIEW

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/connector/protocol/thrift/SyncClientManager.java [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+    TEndPoint endPoint = leaderCacheManager.parseEndPoint(event);
+    return endPoint == null
+            || !useLeaderCache
+            || !endPoint2client.containsKey(endPoint)
+            || Boolean.FALSE.equals(endPoint2client.get(endPoint).getRight())
+        ? getOneConnectedClientAndStatus()
+        : endPoint2client.get(endPoint);
+  }
+
+  public Pair<IoTDBThriftSyncConnectorClient, Boolean> getClientAndStatusByEvent(
```
Comment (Caideyipi) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/connector/protocol/thrift/SyncClientManager.java:127]: Functions alike with parameters having the same parent class can be extracted

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/connector/protocol/thrift/sync/IoTDBThriftSyncConnector.java [hunk 2]
```diff
@@ -464,9 +429,21 @@ private void doTransfer(
     }
 
     // 2. Transfer file seal signal, which means the file is transferred completely
-    final TPipeTransferResp resp =
-        client.pipeTransfer(
-            PipeTransferFileSealReq.toTPipeTransferReq(tsFile.getName(), tsFile.length()));
+    final TPipeTransferResp resp;
```
Comment (Caideyipi) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/connector/protocol/thrift/sync/IoTDBThriftSyncConnector.java:432]: May these part of code be extracted to a single method?

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/connector/protocol/thrift/SyncClientManager.java [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+        return clientAndStatus;
+      }
+    }
+    throw new PipeConnectionException(
+        "All clients are dead, please check the connection to the receiver.");
+  }
+
+  public Pair<IoTDBThriftSyncConnectorClient, Boolean> getClientAndStatusByEvent(
+      PipeInsertNodeTabletInsertionEvent event) {
+    TEndPoint endPoint = leaderCacheManager.parseEndPoint(event);
```
Comment (Caideyipi) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/connector/protocol/thrift/SyncClientManager.java:118]: The name of the "endpoint" can be a little more explanatory, such as "leaderEndpoint".

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/connector/protocol/thrift/SyncClientManager.java [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+      if (Boolean.TRUE.equals(endPoint2client.get(nodeUrl).getRight())) {
+        return;
+      }
+    }
+
+    throw new PipeConnectionException(
+        String.format("All target servers %s are not available.", endPoint2client.keySet()));
+  }
+
+  public Pair<IoTDBThriftSyncConnectorClient, Boolean> getOneConnectedClientAndStatus() {
```
Comment (Caideyipi) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/connector/protocol/thrift/SyncClientManager.java:101]: Maybe "getNext" is better?

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/connector/protocol/thrift/SyncClientManager.java [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+    }
+  }
+
+  public void clean() {
+    for (TEndPoint endPoint : endPoint2client.keySet()) {
+      try {
+        if (endPoint2client.get(endPoint).getLeft() != null) {
+          endPoint2client.get(endPoint).getLeft().close();
+          endPoint2client.get(endPoint).setLeft(null);
+          endPoint2client.get(endPoint).setRight(null);
```
Comment (Caideyipi) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/connector/protocol/thrift/SyncClientManager.java:165]: This "null" set to status may be redundant?

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/connector/protocol/thrift/LeaderCacheManager.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    return insertNode == null
+        ? null
+        : getEndPointByDeviceId(insertNode.getDevicePath().getFullPath());
+  }
+
+  public TEndPoint parseEndPoint(PipeRawTabletInsertionEvent event) {
+    return getEndPointByDeviceId(event.getDeviceId());
+  }
+
+  public void updateOrCreate(String deviceId, TEndPoint endPoint) {
```
Comment (Caideyipi) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/connector/protocol/thrift/LeaderCacheManager.java:48]: What about directly put it?

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/connector/protocol/thrift/SyncClientManager.java [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+  }
+
+  public Pair<IoTDBThriftSyncConnectorClient, Boolean> getClientAndStatusByEvent(
+      PipeTsFileInsertionEvent event) {
+    return getOneConnectedClientAndStatus();
+  }
+
+  public void updateOrCreate(String deviceId, TEndPoint endPoint) {
+    try {
+      endPoints.add(endPoint);
```
Comment (Caideyipi) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/connector/protocol/thrift/SyncClientManager.java:145]: The "endPoints" may be useless after the first call to "initClients()", by the way, if you "add" an endpoint to update that you may be expecting the "endPoints" to be a set.

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/connector/protocol/thrift/SyncClientManager.java [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+        LOGGER.warn(
+            "Failed to close client {}:{}, because: {}.", endPoint.getIp(), endPoint.getPort(), e);
+      } finally {
+        endPoint2client.get(endPoint).setRight(false);
+      }
+    }
+  }
+
+  private void initAndSetClient(TEndPoint endPoint) throws TTransportException, IOException {
+    if (endPoint2client.get(endPoint).getLeft() != null) {
```
Comment (Caideyipi) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/connector/protocol/thrift/SyncClientManager.java:177]: Closing a client upon redirection is a little bit not graceful...... Besides, if a client is killed here, the status will not be set to false, and if it does not start successfully....

**Review by Caideyipi — CHANGES_REQUESTED** [review]
PTAL....

**Review by SteveYurongSu — APPROVED** [review]

**Pull request merged**